app/src/main/java/com/aaronjwood/portauthority/async/ScanHostsAsyncTask.java

```diff
@@ -37,26 +37,26 @@ public ScanHostsAsyncTask(MainAsyncResponse delegate) {
      * @param params IP address
      */
     @Override
-    protected Void doInBackground(String... params) {
-        String ip = params[0];
-        String parts[] = ip.split("\\.");
+    protected Void doInBackground(Integer... params) {
+        int ipv4 = params[0];
+        int cidr = params[1];
 
         ExecutorService executor = Executors.newCachedThreadPool();
 
+        double hostBits = 32.0d - cidr; // How many bits do we have for the hosts.
+        int netmask = (0xffffffff >> (32 - cidr)) << (32 - cidr); // How many bits for the netmask.
+        int numberOfHosts = (int) Math.pow(2.0d, hostBits) - 2; // 2 ^ hostbits = number of hosts in integer.
+        int firstAddr = (ipv4 & netmask) + 1; // AND the bits we care about, then first addr.
+
         int SCAN_THREADS = 8;
-        int chunk = (int) Math.ceil((double) 255 / SCAN_THREADS);
-        int previousStart = 1;
-        int previousStop = chunk;
+        int chunk = (int) Math.ceil((double) numberOfHosts / SCAN_THREADS); // Chunk hosts by number of threads.
+        int previousStart = firstAddr;
+        int previousStop = firstAddr + (chunk - 2); // Ignore network + first addr
 
         for (int i = 0; i < SCAN_THREADS; i++) {
-            if (previousStop >= 255) {
-                previousStop = 255;
-                executor.execute(new ScanHostsRunnable(parts, previousStart, previousStop, delegate));
-                break;
-            }
-            executor.execute(new ScanHostsRunnable(parts, previousStart, previousStop, delegate));
+            executor.execute(new ScanHostsRunnable(previousStart, previousStop, delegate));
```

Comment:
Don't you need some logic to check for how the rounding and `previousStop` is maintained? For example, if `chunk` does not evenly go into `numberOfHosts` so you'll end up going over the range in the last iteration of the loop.